src/workerd/io/worker-interface.capnp

```diff
@@ -157,6 +157,125 @@ struct Trace @0x8e8d911203762d34 {
+    # the fetch handler returns a Response. Importantly, it does not signal that the
+    # span has closed, which may not happen for some period of time after the return
+    # mark is recorded (e.g. due to things like waitUntils or waiting to fully ready
+    # the response body payload, etc).
+    info :union {
+      empty @0 :Void;
+      custom @1 :List(Attribute);
+      fetch @2 :FetchResponseInfo;
+    }
+  }
+
+  struct SpanOpen {
+    # Marks the opening of a child span within the streaming tail session.
+    operationName @0 :Text;
+    info :union {
+      empty @1 :Void;
+      custom @2 :List(Attribute);
+      fetch @3 :FetchEventInfo;
+      jsrpc @4 :JsRpcEventInfo;
+    }
+  }
+
+  struct SpanClose {
+    # Marks the closing of a child span within the streaming tail session.
+    # Once emitted, no further mark events should occur within the closed
+    # span.
+    outcome @0 :EventOutcome;
+  }
+
+  struct Resume {
+    # A resume event indicates that we are resuming a previously hibernated
+    # tail session.
+    attachment @0 :Data;
+  }
+
+  struct Onset {
+    # The Onset and Outcome event types are special forms of SpanOpen and
+    # SpanClose that explicitly mark the start and end of the root span.
+    # A streaming tail session will always begin with an Onset event, and
+    # always end with an Outcome event.
+    executionModel @0 :ExecutionModel;
+    scriptName @1 :Text;
+    scriptVersion @2 :ScriptVersion;
+    dispatchNamespace @3 :Text;
+    scriptTags @4 :List(Text);
+    entryPoint @5 :Text;
+
+    info :union {
+      fetch @6 :FetchEventInfo;
+      jsrpc @7 :JsRpcEventInfo;
+      scheduled @8 :ScheduledEventInfo;
+      alarm @9 :AlarmEventInfo;
+      queue @10 :QueueEventInfo;
+      email @11 :EmailEventInfo;
+      trace @12 :TraceEventInfo;
+      hibernatableWebSocket @13 :HibernatableWebSocketEventInfo;
+      resume @14 :Resume;
+      custom @15 :CustomEventInfo;
+    }
+  }
+
+  struct Outcome {
+    outcome @0 :EventOutcome;
+    cpuTime @1 :UInt64;
+    wallTime @2 :UInt64;
+  }
+
+  struct Hibernate {
+    # A hibernate event indicates that the tail session is being hibernated.
+  }
+
+  struct TailEvent {
+    # A streaming tail worker receives a series of Tail Events. Tail events always
+    # occur within an InvocationSpanContext. The first TailEvent delivered to a
+    # streaming tail session is always an Onset. The final TailEvent delivered is
+    # always an Outcome. Between those can be any number of SpanOpen, SpanClose,
+    # and Mark events. Every SpanOpen *must* be associated with a SpanClose unless
+    # the stream was abruptly terminated.
+    context @0 :InvocationSpanContext;
+    parentContext @1 :InvocationSpanContext;
+    timestampNs @2 :Int64;
+    sequence @3 :UInt32;
+    event :union {
+      onset @4 :Onset;
+      outcome @5 :Outcome;
+      hibernate @6 :Hibernate;
+      spanOpen @7 :SpanOpen;
+      spanClose @8 :SpanClose;
+      attribute @9 :Attribute;
```

Comment:
Can a single attribute be an event? From the I assumed that attributes would always have an associated span and all attributes of a span would be reported together in the Return event.

Reply:
Setting a span attribute will be an event, yes. They have nothing to do with the `Return` event. All `Attribute` events received within a span are attributes of that span.

Reply:
Having every tag be a separate event feels rather inefficient – if a span has say 3 - 6 small tags (or attributes in OTel terminology), which will apply for several user tracing span types, that means having many more events to be sent even though the tags are mostly being set all at the same time.
For custom spans it is conceivable that sometimes tags will be created at different points in time within the span, and that the tail worker relies on getting that tag information before the span is complete (this is kind of at odds with how I think span tags should be used but it's plausible).
To avoid the overhead, would adding an attributes list to the SpanEnd event make sense so that for user tracing tags are part of that event and not distinct events? When we support custom spans in the future we can add support for manually adding tags that will be their own events.

Reply:
Yep, for this case I've been considering making it so that the `Attribute` event, multiple attributes can actually be provided at once.